File: Ridy-taxi-app/ios/common/Models/Objects/Request.swift
//
//  Travel.swift
//
//  Copyright (c) Minimalistic Apps. All rights reserved.
//

import Foundation
import CoreLocation

public final class Request: Codable, Hashable {
    public static func == (lhs: Request, rhs: Request) -> Bool {
        return lhs.id == rhs.id
    }
    
    public static var shared = Request()
    
    // MARK: Properties
    public var addresses: [String] = []
    public var points: [CLLocationCoordinate2D] = []
    public var durationReal: Int?
    public var distanceReal: Int?
    public var requestTimestamp: Double?
    public var expectedTimestamp: Double?
    public var currency: String?
    public var startTimestamp: Double?
    public var etaPickup: Double?
    public var finishTimestamp: Double?
    public var driver: Driver?
    public var coupon: Coupon?
    public var rating: Int?
    public var costBest: Double?
    public var costAfterCoupon: Double?
    public var providerShare: Double?
    public var rider: Rider?
    public var distanceBest: Int?
    public var status: Status?
    public var durationBest: Int?
    public var cost: Double?
    public var id: Int?
    public var isHidden: Int?
    public var log: String?
    public var service: Service?
    public var confirmationCode: Int?
    
    public enum Status: String, Codable {
        case Requested = "Requested"
        case NotFound = "NotFound"
        case NoCloseFound = "NoCloseFound"
        case Found = "Found"
        case DriverAccepted = "DriverAccepted"
        case WaitingForPrePay = "WaitingForPrePay"
        case DriverCanceled = "DriverCanceled"
        case RiderCanceled = "RiderCanceled"
        case Arrived = "Arrived"
        case Started = "Started"
        case WaitingForPostPay = "WaitingForPostPay"
        case WaitingForReview = "WaitingForReview"
        case Finished = "Finished"
        case Booked = "Booked"
        case Expired = "Expired"
    }
    
    public func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

File: Ridy-taxi-app/android/common/src/main/java/com/innomalist/taxi/common/models/Request.kt
package com.innomalist.taxi.common.models

import com.google.android.gms.maps.model.LatLng
import com.innomalist.taxi.common.R
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class Request(
        var driver: Driver? = null,
        var rider: Rider? = null,
        var cost: Double? = null,
        var startTimestamp: Long? = 0,
        var log: String? = null,
        var distanceBest: Int? = null,
        var rating: Int? = null,
        var isHidden: Int? = null,
        var addresses: List<String> = ArrayList(),
        var points: List<LatLng> = ArrayList(),
        var finishTimestamp: Long? = null,
        var requestTimestamp: Long? = null,
        var etaPickup: Long? = 0,
        var durationBest: Int? = null,
        var costBest: Double? = null,
        var costAfterCoupon: Double? = null,
        var providerShare: Double? = 0.0,
        var currency: String? = null,
        var durationReal: Long? = null,
        var distanceReal: Long? = null,
        var id: Long? = null,
        var status: Status? = null,
        var imageUrl: String? = null,
        var service: Service? = null,
        var confirmationCode: Int? = null
) {
    enum class Status {
        @Json(name = "Requested")
        Requested,
        @Json(name = "NotFound")
        NotFound,
        @Json(name = "NoCloseFound")
        NoCloseFound,
        @Json(name = "Found")
        Found,
        @Json(name = "DriverAccepted")
        DriverAccepted,
        @Json(name = "Arrived")
        Arrived,
        @Json(name = "WaitingForPrePay")
        WaitingForPrePay,
        @Json(name = "RiderCanceled")
        RiderCanceled,
        @Json(name = "DriverCanceled")
        DriverCanceled,
        @Json(name = "WaitingForPostPay")
        WaitingForPostPay,
        @Json(name = "WaitingForReview")
        WaitingForReview,
        @Json(name = "Started")
        Started,
        @Json(name = "Booked")
        Booked,
        @Json(name = "Expired")
        Expired,
        @Json(name = "Finished")
        Finished;

        val localizedDescription: Int
            get() {
                return when (this) {
                    Requested -> R.string.request_status_requested
                    NotFound -> R.string.request_status_not_found
                    NoCloseFound -> R.string.request_status_no_close_found
                    Found -> R.string.request_status_found
                    DriverAccepted -> R.string.request_status_driver_accepted
                    Arrived -> R.string.request_status_arrived
                    WaitingForPrePay -> R.string.request_status_waiting_for_pre_pay
                    RiderCanceled -> R.string.request_status_rider_canceled
                    DriverCanceled -> R.string.request_status_driver_canceled
                    WaitingForPostPay -> R.string.request_status_waiting_for_post_pay
                    WaitingForReview -> R.string.request_status_waiting_for_review
                    Started -> R.string.request_status_started
                    Booked -> R.string.request_status_booked
                    Expired -> R.string.request_status_expired
                    Finished -> R.string.request_status_finished
                }
            }
    }

}


